### ISSUE TITLE
maps: add Merge, MergeFunc, Filter funcs

### ISSUE THREAD

**Issue by idsulik** [golang/go#68480]
I want to propose adding new functions to the maps package:

```
// Merge adds all key/value pairs in m2 to m1.
func Merge[M1 ~map[K]V, M2 ~map[K]V, K comparable, V any](m1 M1, m2 M2) {
	for k, v := range m2 {
		m1[k] = v
	}
}

// MergeFunc is like Merge, but calls the merge function to resolve duplicate keys.
func MergeFunc[M1 ~map[K]V, M2 ~map[K]V, K comparable, V any](m1 M1, m2 M2, merge func(K, V, V) V) {
	for k, v := range m2 {
		if v1, ok := m1[k]; ok {
			m1[k] = merge(k, v1, v)
		} else {
			m1[k] = v
		}
	}
}

// Filter creates a new map containing only the key-value pairs that match a predicate
func Filter[M ~map[K]V, K comparable, V any](m M, keep func(K, V) bool) M {
	m2 := make(M, len(m))
	for k, v := range m {
		if keep(k, v) {
			m2[k] = v
		}
	}
	return m2
}
```

I'll be happy to contribute a PR with tests if this is a good idea.

**Comment by gophun** [comment]
How is Merge different from [Copy](https://pkg.go.dev/maps#Copy)?

**Comment by idsulik** [comment]
> How is Merge different from [Copy](https://pkg.go.dev/maps#Copy)?

You're right, they're the same, I think we can keep MergeFunc, but replace it with CopyFunc so we can resolve key conflicts

**Comment by jimmyfrasche** [comment]
See the discussion for a generic set type: https://github.com/golang/go/discussions/47331 (possibly to be revived soon once iterators land in the next version, as that's what stalled the original discussion)

I also wrote https://pkg.go.dev/github.com/jimmyfrasche/mapset for doing set like things with regular maps

**Comment by ianlancetaylor** [comment]
`Filter` seems like something for #61898.  I don't see why we need a maps-specific filter.

**Comment by idsulik** [comment]
> Filter seems like something for https://github.com/golang/go/issues/61898. I don't see why we need a maps-specific filter.
The most important and useful are `Diff` and `Intersect`.  
Filter will be useful for example, if you create map of files using some external package and you need to exclude some files from the map, or you create some items counter map and need to exclude items with less than 10 count

**Comment by ianlancetaylor** [comment]
I'm not saying that `Filter` is not useful.  I'm saying that the `Filter2` from #61898 is fine, and we don't also need a separate `maps.Filter`.

**Comment by isgj** [comment]
> The most important and useful are Diff and Intersect.

The first thing that I thought when reading the title was how will you compare the values.
I think the documentation should mention that only the keys of the maps are used for comparison. IMHO these are not natural operations of maps.

**Comment by idsulik** [comment]
> I'm not saying that `Filter` is not useful. I'm saying that the `Filter2` from #61898 is fine, and we don't also need a separate `maps.Filter`.

the filter allows you to filter only based on value, this one will also provide you `key`

**Comment by idsulik** [comment]
> I think the documentation should mention that only the keys of the maps are used for comparison

We can add value checking as well, if key is not found or values are diff, then append it to result map

**Comment by ianlancetaylor** [comment]
> the filter allows you to filter only based on value, this one will also provide you `key`

`Filter2` filters based on both value and key.

**Comment by idsulik** [comment]
> `Filter2` filters based on both value and key.

You're right, I missed it. so I think I can implement Diff and Intersect

**Comment by isgj** [comment]
> We can add value checking as well, if key is not found or values are diff

How will you compare the values? Or will you restrict the type constraint of `V` to `comparable`?

**Comment by idsulik** [comment]
> Or will you restrict the type constraint of `V` to `comparable`?

I see only this way for this
